### ISSUE TITLE
AIX support

### ISSUE THREAD

**Issue by koldat** [facebook/rocksdb#966]
Hi,

I built RocksDB on AIX including JNI. It required many changes in sources and also has prerequisite for my other pull requests. I will wait until others will be merged until I will create another pull request.

I just created issue to track the progress and discussion. First of all I would like to ask if RocksDB team wants to merge AIX build?

Thanks,
Tomas

**Comment by mdcallag** [comment]
Wow, I hope this gets done. How will we avoid breaking AIX support going forward? OSUOSL has Power servers for open source projects. Not sure who is signed up to manage that.
http://osuosl.org/services/powerdev

**Comment by koldat** [comment]
You are right. Continuous integration would be nice. We will need to investigate our options. Right now I am using my company resources (CA Technologies) as part of our project. It is quite hard to prepare the machine and I am not sure about all AIX version we can support. I was successful on AIX 7.1 ppc64 using gcc 4.8.3 based on 4.1.fb branch. But it is a good start. I will try to use official AIX compiler xlc, but it will probably have problems with latest c++11 that is used on many places.

I just want to push initial support into master (most of the tests are passing - 95%) that is not affecting others so that I will not need to do merges all the time. After some time we can agree on some build automation.

**Comment by dhruba** [comment]
@koldat : curious to know if you were able to run RocksDB on AIX?

**Comment by koldat** [comment]
@dhruba Yes we make it working. I have JNI on branch 4.1 running on AIX, Solaris Sparc, Mac, Windows, Linux. All Java tests passes. Very small number do not pass, because of problem with endianity (cuckoo hashing). We are finalizing tests and then I will start merge to master branch so I can do pull request. Hopefully soon.

**Comment by dhruba** [comment]
Thanks for the update @koldat.

**Comment by dhruba** [comment]
@koldat If you have any patches to contribute back to open-source rocksdb, that would be great.

**Comment by koldat** [comment]
Sorry that we did not do it yet. We were finalizing our product and fixing some of the stuff related to RocksDB. I will plan this work for next sprint we will have.

**Comment by manishbansal8843** [comment]
@koldat Any progress on this? I am also trying to build rocksdb jni for one of my kafka streaming application. I am using AIX 6.1 and  g++ 4.8.5 to build the same. However, i am facing lot of issues. You can check issue #2091 . It would be great if you can provide any help here.

**Comment by koldat** [comment]
We have ported changes to RocksDB5.1, because Rocks moved a lot from 4.1. It was not possible to merge. Unfortunately we have our own build for it to support all 5 platforms in one build easily (more Maven way). We have to separate the changes. I can push everything into my space is you are interested, but it is not "pull-requestable". We have to pull request the separated one.

And yes it works on AIX6.1 gcc 4.8.5.

**Comment by manishbansal8843** [comment]
@koldat ohh great if it works on AIX 6.1. Yes kindly push it. Any kind of help would be appreciated.

**Comment by koldat** [comment]
I did some rebase and merged master. Right now I am trying to make all platforms working (there was quite refactoring in master). I am almost there. You can take a look here for current state: 
https://github.com/koldat/rocksdb/tree/aix_support_master

I think I will fix the rest tomorrow.

**Comment by manishbansal8843** [comment]
@koldat Thanks for sharing it. I have tried rebuilding it. However, i am getting link errors as below.

```
Generating target/apidocs/overview-summary.html...
Generating target/apidocs/help-doc.html...
2 warnings
gmake[1]: Leaving directory '/users2/e462197/rocksdb/koldat/rocksdb-aix_support_master/java'
ld: 0711-317 ERROR: Undefined symbol: .std::basic_ios<char, std::char_traits<char> >::clear(std::_Ios_Iostate)
ld: 0711-317 ERROR: Undefined symbol: typeinfo for std::out_of_range
ld: 0711-317 ERROR: Undefined symbol: .std::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(char const*, std::allocator<char> const&)
ld: 0711-317 ERROR: Undefined symbol: .std::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
..
//some more logs
..
ld: 0711-317 ERROR: Undefined symbol: .std::basic_string<char, std::char_traits<char>, std::allocator<char> >::find_last_of(char const*, unsigned long, unsigned long) const
ld: 0711-345 Use the -bloadmap or -bnoquiet option to obtain more information.
collect2: error: ld returned 8 exit status
Makefile:1562: recipe for target 'rocksdbjava' failed
gmake: *** [rocksdbjava] Error 1
```
Any idea on this? Am i missing some packages to install as a pre-requisite.

**Comment by koldat** [comment]
This is what we are using:

```
sed "s/1.0.0-SNAPSHOT/${BUILD_VERSION}/" java/deploy/aix64/pom.xml > java/deploy/aix64/pom.bak.xml
cp -f java/deploy/aix64/pom.bak.xml java/deploy/aix64/pom.xml
export PORTABLE=1
export CC=gcc
export AR="ar -X64"
export EXTRA_ARFLAGS=-X64
export EXTRA_CFLAGS=-maix64
export EXTRA_CXXFLAGS=-maix64
export PLATFORM_LDFLAGS="-static-libstdc++ -static-libgcc"

export SNAPPY_DOWNLOAD_BASE="-k https://github.com/google/snappy/releases/download"
export LZ4_DOWNLOAD_BASE="-k https://github.com/lz4/lz4/archive"

make clean
make rocksdbjavastatic
```
I am still having some test failures on Linux after merge. I can push 5.1.4 branch for you.

You also need to downgrade AIX assembler processor. Latest releases contains issues. I have:
bash-4.1# installp -l -d . | grep bos.adt.base
  bos.adt.base                6.1.6.15                   S  N usr,root

We are using GCC and all related packages (stdc, ...) in version 4.8.5-1:
http://www.oss4aix.org/download/RPMS/gcc/

**Comment by ayappanec** [comment]
@koldat I actually recommend downloading AIX rpms from the official Toolbox. ftp://public.dhe.ibm.com/aix/freeSoftware/aixtoolbox/RPMS/ .
We actually do lot of testing here and we do have yum available for customers .  ftp://public.dhe.ibm.com/aix/freeSoftware/aixtoolbox/ezinstall/ppc/README-yum

**Comment by koldat** [comment]
@ayappanec maybe, but we are using one I listed and it works perfectly. But thanks for pointer.

**Comment by manishbansal8843** [comment]
@koldat I tried using the instructions you shared. Although i am not able to check AIX assembler processor, still i tried to proceed with it. Also, my machine seems to have some firewall settings due to which on the fly download of various compressors like snappy and zlib etc are not working. For that, i have pre-downloaded them and commented the relevant lines in makefile. Now, i am facing issues while libsnappy.a target is executed. Below are the logs.

```
cd snappy-1.1.4 && CFLAGS='-maix64' CXXFLAGS='-maix64' LDFLAGS='' ./configure --with-pic --enable-static --disable-shared
checking for a BSD-compatible install... ./install-sh -c
checking whether build environment is sane... yes
checking for a thread-safe mkdir -p... ./install-sh -c -d
checking for gawk... no
checking for mawk... no
checking for nawk... nawk
checking whether make sets $(MAKE)... yes
checking whether make supports nested variables... yes
checking build system type... /usr/bin/oslevel[7]: 14614946 Killed
config.sub: too many arguments
Try `config.sub --help' for more information.
configure: error: /bin/sh ./config.sub powerpc-ibm-aixoslevel: Error processing cache, return code 137. oslevel: Use the -f flag to rebuild cache. failed
Makefile:1489: recipe for target 'libsnappy.a' failed
gmake: *** [libsnappy.a] Error 1
```
P.S. target libz.a and libbz2.a got executed successfully,

**Comment by koldat** [comment]
This is my output:
```
00:02:29 cd snappy-1.1.4 && CFLAGS='-maix64' CXXFLAGS='-maix64' LDFLAGS='' ./configure --with-pic --enable-static --disable-shared
00:02:29 checking for a BSD-compatible install... /opt/freeware/bin/install -c
00:02:31 checking whether build environment is sane... yes
00:02:31 checking for a thread-safe mkdir -p... /opt/freeware/bin/mkdir -p
00:02:31 checking for gawk... no
00:02:31 checking for mawk... no
00:02:31 checking for nawk... nawk
00:02:31 checking whether make sets $(MAKE)... yes
00:02:31 checking whether make supports nested variables... yes
00:02:31 checking build system type... powerpc-ibm-aix6.1.0.0
00:02:32 checking host system type... powerpc-ibm-aix6.1.0.0
00:02:32 checking how to print strings... print -r
00:02:32 checking for style of include used by make... GNU
00:02:32 checking for gcc... gcc
00:02:32 checking whether the C compiler works... yes
00:02:33 checking for C compiler default output file name... a.out
00:02:33 checking for suffix of executables... 
```

Anyway you can comment out snappy build. And remove snappy from -D and linker. We were doing the same on beginning. I am not sure what problem you have. You have to look in configure script and log.

**Comment by ayappanec** [comment]
@manishbansal8843 You need to have /opt/freeware/bin in your PATH variable.
export PATH=/opt/freeware/bin:$PATH

**Comment by koldat** [comment]
Thanks @ayappanec , you are right. Here are our Jenkins env overrides:
JAVA_HOME=/usr/jenkins/ibm-java-ppc64-70
LIBPATH=/opt/freeware/lib
PATH=/opt/freeware/bin:/usr/bin:/etc:/usr/sbin:/usr/ucb:/usr/bin/X11:/sbin:/usr/jenkins/ibm-java-ppc64-70/jre/bin:/usr/jenkins/ibm-java-ppc64-70/bin:/usr/lib/apache-maven-3.2.1/bin

**Comment by manishbansal8843** [comment]
@koldat @ayappanec I have added /opt/freeware/lib to both LIBPATH and PATH. However, snappy issue is still coming. Further, after commenting snappy target in makefile, i am again getting the same linking errors as i got yesterday. 

```
ld: 0711-317 ERROR: Undefined symbol: .std::basic_ios<char, std::char_traits<char> >::clear(std::_Ios_Iostate)
ld: 0711-317 ERROR: Undefined symbol: typeinfo for std::out_of_range
ld: 0711-317 ERROR: Undefined symbol: .std::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(char const*, std::allocator<char> const&)
ld: 0711-317 ERROR: Undefined symbol: .std::basic_string<char, std::char_traits<char>, std::allocator<char> >::basic_string(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)
ld: 0711-317 ERROR: Undefined symbol: std::basic_string<char, std::char_traits<char>, std::allocator<char> >::_Rep::_S_empty_rep_storage
ld: 0711-317 ERROR: Undefined symbol: .std::basic_string<char, std::char_traits<char>, std::allocator<char> >::_Rep::_M_destroy(std::allocator<char> const&)

//Some more logs

ld: 0711-317 ERROR: Undefined symbol: .std::basic_istringstream<char, std::char_traits<char>, std::allocator<char> >::basic_istringstream(std::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, std::_Ios_Openmode)
ld: 0711-317 ERROR: Undefined symbol: .std::basic_istream<char, std::char_traits<char> >& std::basic_istream<char, std::char_traits<char> >::_M_extract<unsigned long>(unsigned long&)
ld: 0711-317 ERROR: Undefined symbol: .std::basic_string<char, std::char_traits<char>, std::allocator<char> >::find_last_of(char const*, unsigned long, unsigned long) const
ld: 0711-345 Use the -bloadmap or -bnoquiet option to obtain more information.
collect2: error: ld returned 8 exit status
Makefile:1526: recipe for target 'rocksdbjavastatic' failed
gmake: *** [rocksdbjavastatic] Error 1
```

**Comment by koldat** [comment]
Fixes for AIX was merged in #2194
